url0e7d57473a_kyorohiro_HetimaUtil_tgz-pJ8-net_hetimatan_net_http_HttpServer3xxJ8/src/net/hetimatan/net/http/request/HttpGetRequester.java

```diff
@@ -0,0 +1,141 @@
+	private MessageSendTask mTask = null;
+	private CashKyoroFile mSendCash = null;
+
+	// --- <omake> ---
+	private boolean mIsPostMode = false;
+	public void setBody(KyoroFile body) {
+		mBuilder.putBody(body);
+	}
+	public void isPostMode(boolean on) {
+		mIsPostMode = on;
+	}
+	public boolean isPostMode() {
+		return mIsPostMode;
+	}
+	// --- </omake> ---
+
+	public HttpGetRequester() {
+	}
+
+	public static void log(String message) {
+		System.out.println("KyoroSocketGetRequester#"  + message);
+	}
+
+	public HttpRequestBuilder getUrlBuilder() {
+		return mBuilder;
+	}
+
+	public KyoroSocket connect(KyoroSocket socket) throws IOException, InterruptedException {
+		if(socket == null) {
+			socket = new KyoroSocketImpl();
+		}
+		socket.setDebug("KyoroSocketGetConnection:"+mBuilder.getHost() +","+ mBuilder.getPort());
+		socket.connect(mBuilder.getHost(), mBuilder.getPort());
+		return socket;
+	}
+
+	public MessageSendTask getRequestTask(KyoroSocket socket) throws IOException {
+		log("_writeRequest()");
+		byte[] buffer = createRequest();
+		log(new String(buffer));
+		if(mSendCash == null) {
+			mSendCash = new CashKyoroFile(1024, 3);
+		}
+		if(mTask == null) {
+			mSendCash.addChunk(buffer);
+			mTask = new MessageSendTask(socket, mSendCash);
+		}
+		return mTask;
+	}
+
+	public HttpGetResponse getResponse(KyoroSocket socket) throws IOException, InterruptedException {
+		HttpGetResponse response = new HttpGetResponse(socket);
+		return response;
+	}	
+
+	public synchronized HttpRequest createHttpRequest() throws IOException {
+		if(false == mIsPostMode) {
+			return mBuilder.createHttpGetRequest();
+		} else {
+			// omake
+			return mBuilder.createHttpPostRequest();
+		}
+	}
+
+	public synchronized byte[] createRequest() throws IOException {
+		HttpRequest uri = createHttpRequest();
+		return HttpObject.createEncode(uri).getBytes();
+	}
+
+	public void close() {
+		try {
+			if(mSendCash != null) {
+				mSendCash.close();
+			}
+		} catch(IOException e) {
+
+		}
+	}
+	/**
+	 * テスト用
+	 * @throws Throwable 
+	 */
+	public static HttpGetResponse syncRequest(HttpGetRequester requester, KyoroSelector selector) throws Throwable {
+		KyoroSocket socket = null;
+		try {
+			socket = requester.connect(null);
+			while(socket.getConnectionState() == KyoroSocket.CN_CONNECTING){
+				Thread.yield();
+				Thread.sleep(0);}
+			MessageSendTask sendTask= requester.getRequestTask(socket);
+			do {
+				sendTask.action(null);
+			} while(sendTask.isKeep());
+			
+			//
+			HttpGetResponse res = requester.getResponse(socket);
+			while(!res.headerIsReadable()){;}
+			res.readHeader();
+			while(!res.bodyIsReadable()){;}
+			res.readBody();
+			return res;
+		} finally {
+			if (socket != null) {
+				socket.close();
+			}
+		}
+	}
+
+	/**
+	 * テスト用
+	 * @throws Throwable 
+	 */
+	public static HttpGetResponse doRequest(HttpGetRequester requester) throws Throwable {
+		return syncRequest(requester, new KyoroSelector());
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
Analysis of this package determined that this line of code might contain a resource leak. A resource leak can slow down or crash your system.

**Fix**
Consider closing the resource returned by the *KyoroSelector* constructor in a try-finally block. Execution paths that do not contain closure statements were detected.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).